src/libs/ajax.ts

```diff
@@ -0,0 +1,243 @@
+                    methodUri: `agora://${namespace}/${name}/${snapshotId}`,
+                  },
+                  name,
+                  namespace,
+                  rootEntityType: '',
+                  prerequisites: {},
+                  inputs: {},
+                  outputs: {},
+                  methodConfigVersion: 1,
+                  deleted: false,
+                },
+                config.payloadObject
+              )
+            ),
+            { signal, method: 'POST' },
+          ])
+        );
+        return res.json();
+      },
+    };
+  },
+});
+
+const Submissions = (signal?: AbortSignal) => ({
+  queueStatus: async () => {
+    const res = await fetchRawls('submissions/queueStatus', _.merge(authOpts(), { signal }));
+    return res.json();
+  },
+});
+
+const DrsUriResolver = (signal?: AbortSignal) => ({
+  // DRSHub now gets a signed URL instead of Martha
+  getSignedUrl: async ({ bucket, object, dataObjectUri, googleProject }) => {
+    const res = await fetchDrsHub(
+      '/api/v4/gcs/getSignedUrl',
+      _.mergeAll([
+        jsonBody({ bucket, object, dataObjectUri, googleProject }),
+        authOpts(),
+        appIdentifier,
+        { signal, method: 'POST' },
+      ])
+    );
+    return res.json();
+  },
+
+  getDataObjectMetadata: async (url, fields) => {
+    const res = await fetchDrsHub(
+      '/api/v4/drs/resolve',
+      _.mergeAll([jsonBody({ url, fields }), authOpts(), appIdentifier, { signal, method: 'POST' }])
+    );
+    return res.json();
+  },
+});
+
+const Surveys = (signal?: AbortSignal) => ({
+  submitForm: (formId, data) => fetchGoogleForms(`${formId}/formResponse?${qs.stringify(data)}`, { signal }),
+});
+
+export const Ajax = (signal?: AbortSignal) => {
+  return {
+    Apps: Apps(signal),
+    AzureStorage: AzureStorage(signal),
+    Billing: Billing(signal),
+    Buckets: GoogleStorage(signal),
+    Catalog: Catalog(signal),
+    Cbas: Cbas(signal),
+    CromIAM: CromIAM(signal),
+    CromwellApp: CromwellApp(signal),
+    DataRepo: DataRepo(signal),
+    Disks: Disks(signal),
+    Dockstore: Dockstore(signal),
+    DrsUriResolver: DrsUriResolver(signal),
+    ExternalCredentials: ExternalCredentials(signal),
+    FirecloudBucket: FirecloudBucket(signal),
+    Groups: Groups(signal),
+    Methods: Methods(signal),
+    Metrics: Metrics(signal),
+    OAuth2: OAuth2(signal),
+    Runtimes: Runtimes(signal),
+    SamResources: SamResources(signal),
+    Submissions: Submissions(signal),
+    Support: Support(signal),
+    Surveys: Surveys(signal),
+    TermsOfService: TermsOfService(signal),
+    User: User(signal),
+    WorkflowScript: WorkflowScript(signal),
+    WorkspaceData: WorkspaceData(signal),
+    WorkspaceManagerResources: WorkspaceManagerResources(signal),
+    Workspaces: Workspaces(signal),
+  };
+};
+
+export type AjaxContract = ReturnType<typeof Ajax>;
+
+// Exposing Ajax for use by integration tests (and debugging, or whatever)
+(window as any).Ajax = Ajax;
```

Comment:
I'm surprised there's not a more specific typing for `window`. Do you think it would be worth trying to type this more specifically, or not worth it. (maybe `window as {} extends Window` or something?)

Reply:
as {} is pretty similar to any.... we are attaching an arbitrary "global" to window, so we need a fairly wiggly type cast.